File: appkit/text_field.m
#import <AppKit/AppKit.h>
#import "text_field.h"
#import "text_field_delegate.h"

@implementation GoNSTextFieldDelegate

- (void)controlTextDidChange:(NSNotification*)notification {
	return TextField_Delegate_ControlTextDidChange([self goID], notification);
}

- (void)controlTextDidEndEditing:(NSNotification*)notification {
	return TextField_Delegate_ControlTextDidEndEditing([self goID], notification);
}

- (void)controlTextDidBeginEditing:(NSNotification*)notification {
	return TextField_Delegate_ControlTextDidBeginEditing([self goID], notification);
}

@end

@interface MyTextField : NSTextField {
}
@end

@implementation MyTextField

- (BOOL)performKeyEquivalent:(NSEvent *)event {
    if ([event type] == NSEventTypeKeyDown) {
        if (([event modifierFlags] & NSEventModifierFlagDeviceIndependentFlagsMask) == NSEventModifierFlagCommand) {
            if ([[event charactersIgnoringModifiers] isEqualToString:@"x"]) {
                if ([NSApp sendAction:@selector(cut:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"c"]) {
                if ([NSApp sendAction:@selector(copy:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"v"]) {
                if ([NSApp sendAction:@selector(paste:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"z"]) {
                if ([NSApp sendAction:@selector(undo:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"a"]) {
                if ([NSApp sendAction:@selector(selectAll:) to:nil from:self]) {
                    return YES;
                }
            }
        } else if (([event modifierFlags] & NSEventModifierFlagDeviceIndependentFlagsMask) == (NSEventModifierFlagCommand | NSEventModifierFlagShift)) {
            if ([[event charactersIgnoringModifiers] isEqualToString:@"Z"]) {
                if ([NSApp sendAction:@selector(redo:) to:nil from:self]) {
                    return YES;
                }
            }
        }
    }
    return [super performKeyEquivalent:event];
}
@end
void TextField_RegisterDelegate(void *ptr, long goID) {
	NSTextField* textField = (NSTextField*)ptr;
	GoNSTextFieldDelegate* delegate = [[GoNSTextFieldDelegate alloc] init];
	[delegate setGoID:goID];
	[textField setDelegate:delegate];
}

bool TextField_IsBezeled(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField isBezeled];
}

void TextField_SetBezeled(void* ptr, bool bezeled) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setBezeled:bezeled];
}

bool TextField_DrawsBackground(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField drawsBackground];
}

void TextField_SetDrawsBackground(void* ptr, bool drawsBackground) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setDrawsBackground:drawsBackground];
}

bool TextField_IsEditable(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField isEditable];
}

void TextField_SetEditable(void* ptr, bool editable) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setEditable:editable];
}

bool TextField_IsSelectable(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField isSelectable];
}

void TextField_SetSelectable(void* ptr, bool selectable) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setSelectable:selectable];
}

void* TextField_TextColor(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField textColor];
}

void TextField_SetTextColor(void* ptr, void* textColor) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setTextColor:(NSColor*)textColor];
}

void* TextField_BackgroundColor(void* ptr) {
	NSTextField* textField = (NSTextField*)ptr;
	return [textField backgroundColor];
}

void TextField_SetBackgroundColor(void* ptr, void* backgroundColor) {
	NSTextField* textField = (NSTextField*)ptr;
	[textField setBackgroundColor:(NSColor*)backgroundColor];
}

void* TextField_NewTextField(NSRect frame) {
	NSTextField* textField = [MyTextField alloc];
	return [[textField initWithFrame:frame] autorelease];
}

File: scripts/text_field.py
#!env python3

from generate import Component, Property, init_method, Param, Method, Return, DelegateMethod


if __name__ == "__main__":
    w = Component(
        Type="appkit.TextField",
        super_type='appkit.Control',
        description="text that the user can select or edit and that sends its action message to its target when the user presses the Return key",
        properties=[
            Property(name='bezeled', Type='bool',
                     description='whether the receiver draws a bezeled border around its contents'),
            Property(name='drawsBackground', Type='bool', getter_prefix_is=False, description='whether the receiver’s cell draws its background color behind its text'),
            Property(name='editable', Type='bool', description='whether the user can edit the value in the text field'),
            Property(name='selectable', Type='bool', description='whether the receiver is selectable (but not editable)'),
            Property(name='textColor', Type='appkit.Color', description='the color used to draw the receiver’s text'),
            Property(name='backgroundColor', Type='appkit.Color', description='the color of the background that the receiver’s cell draws behind the text.'),
        ],
        methods=[ 
            init_method(name="initWithFrame", Type="appkit.TextField", params=[Param(name='frame', Type='foundation.Rect')]),
        ],
        delegate_methods=[
            DelegateMethod(
                name='controlTextDidChange',
                params=[Param(name='notification', Type='foundation.Notification')],
                description=''
            ),
            DelegateMethod(
                name='controlTextDidEndEditing',
                params=[Param(name='notification', Type='foundation.Notification')],
                description=''
            ),
            DelegateMethod(
                name='controlTextDidBeginEditing',
                params=[Param(name='notification', Type='foundation.Notification')],
                description=''
            ),
        ],
        override_methods=[
'''- (BOOL)performKeyEquivalent:(NSEvent *)event {
    if ([event type] == NSEventTypeKeyDown) {
        if (([event modifierFlags] & NSEventModifierFlagDeviceIndependentFlagsMask) == NSEventModifierFlagCommand) {
            if ([[event charactersIgnoringModifiers] isEqualToString:@"x"]) {
                if ([NSApp sendAction:@selector(cut:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"c"]) {
                if ([NSApp sendAction:@selector(copy:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"v"]) {
                if ([NSApp sendAction:@selector(paste:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"z"]) {
                if ([NSApp sendAction:@selector(undo:) to:nil from:self]) {
                    return YES;
                }
            } else if ([[event charactersIgnoringModifiers] isEqualToString:@"a"]) {
                if ([NSApp sendAction:@selector(selectAll:) to:nil from:self]) {
                    return YES;
                }
            }
        } else if (([event modifierFlags] & NSEventModifierFlagDeviceIndependentFlagsMask) == (NSEventModifierFlagCommand | NSEventModifierFlagShift)) {
            if ([[event charactersIgnoringModifiers] isEqualToString:@"Z"]) {
                if ([NSApp sendAction:@selector(redo:) to:nil from:self]) {
                    return YES;
                }
            }
        }
    }
    return [super performKeyEquivalent:event];
}'''
        ]
    )
    w.generate_code()
